ENIGMAsystem/SHELL/Platforms/General/POSIX/POSIXfilemanip.cpp

```diff
@@ -19,51 +19,103 @@
 
 #include <cstdio>
 #include <cstdlib>
-#include <unistd.h>
+#include <string>
+
 #include <sys/stat.h>
+#include <unistd.h>
+#include <dirent.h>
 
-#include <string>
+#define u_root 0
 
-using namespace std;
+using std::string;
 
 /* UNIX-ready port of file manipulation */
 
-namespace enigma_user
-{
+namespace enigma_user {
+namespace {
+static DIR* fff_dir_open = NULL;
+static string fff_mask, fff_path;
+static int fff_attrib;
+}  // namespace
 
-int file_exists(string fname)
-{
+int file_exists(string fname) {
   struct stat st;
-  return (stat(fname.c_str(),&st) == 0) and !(S_ISDIR(st.st_mode));
+  return (stat(fname.c_str(), &st) == 0) and !(S_ISDIR(st.st_mode));
 }
 
-int file_delete(string fname)
-{
-  return remove(fname.c_str());
-}
+int file_delete(string fname) { return remove(fname.c_str()); }
 
-int file_rename(string oldname,string newname)
-{
-  return rename(oldname.c_str(),newname.c_str());
-}
+int file_rename(string oldname, string newname) { return rename(oldname.c_str(), newname.c_str()); }
 
-int file_copy(string fname,string newname)
-{
-  return system(("cp "+fname+" "+newname).c_str()); // Hackish, but there's no good implementation on Linux
+int file_copy(string fname, string newname) {
+  return system(("cp " + fname + " " + newname).c_str());  // Hackish, but there's no good implementation on Linux
 }
 
-int directory_exists(string dname)
-{
+int directory_exists(string dname) {
   struct stat st;
-  return (stat(dname.c_str(),&st) == 0) and (S_ISDIR(st.st_mode));
+  return (stat(dname.c_str(), &st) == 0) and (S_ISDIR(st.st_mode));
 }
 
-int directory_create(string dname) {
-  return mkdir(dname.c_str(),S_IRUSR|S_IWUSR|S_IXUSR);
-}
+int directory_create(string dname) { return mkdir(dname.c_str(), S_IRUSR | S_IWUSR | S_IXUSR); }
+
+int directory_delete(string dname) { return rmdir(dname.c_str()); }
+
+string file_find_next() {
+  if (fff_dir_open == NULL) return "";
+
+  dirent* rd = readdir(fff_dir_open);
+  if (rd == NULL) return "";
+  string r = rd->d_name;
 
-int directory_delete(string dname) {
-  return rmdir(dname.c_str());
+  // Preliminary filter
+
+  const int not_attrib = ~fff_attrib;
+
+  if (r == "." or r == ".."                                                      // Don't return ./ and
+      or ((r[0] == '.' or r[r.length() - 1] == '~') and not_attrib & fa_hidden)  // Filter hidden files
+  )
+    return file_find_next();
+
+  struct stat sb;
+  const string fqfn = fff_path + r;
+  stat(fqfn.c_str(), &sb);
+
+  if ((sb.st_mode & S_IFDIR and not_attrib & fa_directory)          // Filter out/for directories
+      or (sb.st_uid == u_root and not_attrib & fa_sysfile)          // Filter system files
+      or (not_attrib & fa_readonly and access(fqfn.c_str(), W_OK))  // Filter read-only files
+  )
+    return file_find_next();
```

Comment:
This has to be the funniest change in the commit. You added a newline but just left that parenthesis off in space. I don't care at all; I am just surprised you didn't move the parenth to the end of the previous line.